### PR TITLE
TermControl: rework input and focus handling

### LINKED ISSUE #4031: TermControl key down handling investigatio and refactoring
In current TermControl implementation, KeyDown events are handled by PreviewKeyDown event. It uses Tunneling strategy and key down events will first be sent to XAML root element, which is TermControl. 

SearchBoxControl is a child element of TermControl. Thus, key input events on SearchBoxControl will be sent to TermControl first. Now there is an early return check in TermControl::_KeyDownHandler, which prevents key inputs from Search Box to be processed by TermControl. But this is not a good implementation. 

We need to think about using KeyDown instead of PreviewKeyDown event handler in TermControl, but we need to make sure this change won't break anything. 

# Description of the new feature/enhancement

# Proposed technical implementation details (optional)

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
This commit changes how we handle focus and tabstop-ness in TermControl
and its descendent controls.

The root of the issue that #4031 set out to investigate is that
TermControl hosts a bunch of other controls, but it's using event
tunneling to get first dibs on every bit of keyboard input sent its way
-- even for things that live inside it. This required us to hack around
and detemrine whether the input was intended for someone else before
eventually ignoring it or using it.

I believe the _correct_ thing for us to do is to direct user keyboard
input to a control that lives logically "inside" the terminal surface
and have that control hold focus and process keyboard events.

Herein, we introduce an empty and invisible UserControl to hold focus
inside the terminal surface. Everything I've tested works great here.

It's important to note that the UserControl is invisible. Unfortunately,
this means that it can't receive **mouse** input. Its parent, the
SwapChainPanel acting as the terminal surface, can.

I've gone through and unmarked things that should not show up in the
tabstop list and changed the focus parameters for things that should.

Doing this lets us kill TermControl's dependency on all private details
in SearchBoxControl, including the bit where we could ask "do you think
you should be focused?", by letting Xaml handle focus. We now listen to
more focus events and funnel focus where we think it belongs. For
SearchBoxControl, that means we focus the text field and continue to
select all the text in it when the SearchBoxControl gets focus.

Fixes #4031.
This is an alternative solution to #4723, so I reverted that fix.

## PR Checklist
* [x] Closes #4031
* [x] CLA signed
* [x] I've discussed this with core contributors already.

### CODE REVIEW

File: src/cascadia/TerminalControl/SearchBoxControl.xaml [hunk 1]
```diff
@@ -3,15 +3,15 @@
...
     xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
     mc:Ignorable="d"
     x:Name="Root"
     HorizontalAlignment="Stretch"
     VerticalAlignment="Top"
     d:DesignHeight="55"
     d:DesignWidth="285"
-    TabNavigation="Cycle">
+    TabNavigation="Cycle"
+    IsTabStop="True">
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/SearchBoxControl.xaml]: I don't think we actually need `IsTabStop` here. It isn't a tab stop from anything.

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 2]
```diff
@@ -97,20 +96,29 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
+    // Arguments:
+    // - e: the event args for the focus event
+    void TermControl::OnGotFocus(const RoutedEventArgs& e)
+    {
+        if (e.OriginalSource() == *this)
+        {
+            // If we're about to get focus, funnel it into our input site.
+            TerminalInputTarget().Focus(FocusState::Programmatic);
+        }
+    }
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/TermControl.cpp]: So, this means that whenever the TermControl is attempted to get focus, we redirect focus to the TerminalInputTarget right? So you can't ever accidentally give focus to the search box?

Comment (carlos-zamora) [reply src/cascadia/TerminalControl/TermControl.cpp]: Also, just to be clear, this is called just before the `_GotFocusHandler`?

Should we consolidate styles and rename this function to `OnGotFocusHandler`?

File: src/cascadia/TerminalControl/SearchBoxControl.xaml (hunk 1 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalControl/SearchBoxControl.xaml]: "IsTabStop" means "this control, or a control inside it, can receive focus."

File: src/cascadia/TerminalControl/TermControl.cpp (hunk 2 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalControl/TermControl.cpp]: `OnGotFocus` is a special method name in that the UserControl composition class will call it if it's implemented. It must be named this: it is not a handler, it is not bound to an event.

Comment (DHowett-MSFT) [reply src/cascadia/TerminalControl/TermControl.cpp]: If the _TermControl itself_ attempts to receive focus (not the controls inside it, not the search box, not anything else), we redirect focus to the input target.

**Comment by DHowett-MSFT** [comment]
> You changed a few tab stop things. They make sense. You can tab to a SwapChainPanel. But I still can't figure out how to tab to a search box. Honestly, it's ok to just remove the search box as a tab stop because the user can still access it via the search box keybinding.

Information in response to first comment.

File: src/cascadia/TerminalControl/SearchBoxControl.cpp [hunk 3]
```diff
@@ -95,39 +95,19 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
-    // - <none>
-    // Return Value:
-    // - <none>
-    void SearchBoxControl::SetFocusOnTextbox()
+    // - Funnels focus into the textbox and selects it
+    // - Selects all text.
+    void SearchBoxControl::OnGotFocus(/*const IInspectable& sender,*/ Windows::UI::Xaml::RoutedEventArgs const& args)
     {
-        if (TextBox())
+        if (args.OriginalSource() == *this)
```
Comment (leonMSFT) [inline src/cascadia/TerminalControl/SearchBoxControl.cpp]: Just to see if I'm understanding this correctly, this condition returns true _only_ if the `SearchBoxControl` directly gained focus (probably from TermControl)? I'm guessing if the user clicked on the `SearchBoxControl::TextBox()` or its many Toggle-able buttons, the OriginalSource would be the `TextBox` or the buttons, so it wouldn't satisfy this condition?

Comment (DHowett-MSFT) [reply src/cascadia/TerminalControl/SearchBoxControl.cpp]: Indeed! I wanted to maintain the behavior that we had, where the searchbox didn't re-select all its text when the textbox was re-focused, but did when we entered the control for the first time. :smile:

**Comment by miniksa** [comment]
This one's a party of Leon, Mike, and Carlos, so I'm skipping this one intentionally.

**Comment by zadjii-msft** [comment]
Hey I'm gonna mark this as "needs author feedback" - I'm not sure _how_ to resolve the accessibility concerns Carlos had, or if they're still reproing, but maybe we should loop back on this, esp with #6680

**Pull request closed without merging**